tests/utils/utils.go

```diff
@@ -0,0 +1,333 @@
+	if err != nil {
+		return "", 0, fmt.Errorf("failed to parse port: %w", err)
+	}
+
+	return hostAndPort[0], uint32(port), nil
+}
+
+func CreateTeleporterTransaction(
+	ctx context.Context,
+	source SubnetTestInfo,
+	destination SubnetTestInfo,
+	fundedAddress common.Address,
+	fundedKey *ecdsa.PrivateKey,
+	teleporterContractAddress common.Address,
+) *types.Transaction {
+	nonce, err := source.ChainWSClient.NonceAt(ctx, fundedAddress, nil)
+	Expect(err).Should(BeNil())
+
+	data, err := teleporter.EVMTeleporterContractABI.Pack(
+		"sendCrossChainMessage",
+		SendCrossChainMessageInput{
+			DestinationChainID: destination.BlockchainID,
+			DestinationAddress: fundedAddress,
+			FeeInfo: FeeInfo{
+				ContractAddress: fundedAddress,
+				Amount:          big.NewInt(0),
+			},
+			RequiredGasLimit:        big.NewInt(1),
+			AllowedRelayerAddresses: []common.Address{},
+			Message:                 []byte{1, 2, 3, 4},
+		},
+	)
+	Expect(err).Should(BeNil())
+
+	// Send a transaction to the Teleporter contract
+	tx := types.NewTx(&types.DynamicFeeTx{
+		ChainID:   source.ChainIDInt,
+		Nonce:     nonce,
+		To:        &teleporterContractAddress,
+		Gas:       DefaultTeleporterTransactionGas,
+		GasFeeCap: DefaultTeleporterTransactionGasFeeCap,
+		GasTipCap: DefaultTeleporterTransactionGasTipCap,
+		Value:     DefaultTeleporterTransactionValue,
+		Data:      data,
+	})
+
+	txSigner := types.LatestSignerForChainID(source.ChainIDInt)
+	signedTx, err := types.SignTx(tx, txSigner, fundedKey)
+	Expect(err).Should(BeNil())
+
+	return signedTx
+}
+
+// Issues txs to activate the proposer VM fork on the specified subnet index in the manager
+func SetUpProposerVM(ctx context.Context, fundedKey *ecdsa.PrivateKey, manager *runner.NetworkManager, index int) {
+	subnet := manager.GetSubnets()[index]
+	subnetDetails, ok := manager.GetSubnet(subnet)
+	Expect(ok).Should(BeTrue())
+
+	chainID := subnetDetails.BlockchainID
+	uri := HttpToWebsocketURI(subnetDetails.ValidatorURIs[0], chainID.String())
+
+	client, err := ethclient.Dial(uri)
+	Expect(err).Should(BeNil())
+	chainIDInt, err := client.ChainID(ctx)
+	Expect(err).Should(BeNil())
+
+	err = utils.IssueTxsToActivateProposerVMFork(ctx, chainIDInt, fundedKey, client)
+	Expect(err).Should(BeNil())
+}
+
+// Blocks until all validators specified in nodeURIs have reached the specified block height
+func WaitForAllValidatorsToAcceptBlock(ctx context.Context, nodeURIs []string, blockchainID ids.ID, height uint64) {
+	for i, uri := range nodeURIs {
+		chainAWSURI := HttpToWebsocketURI(uri, blockchainID.String())
+		log.Info("Creating ethclient for blockchainA", "wsURI", chainAWSURI)
+		client, err := ethclient.Dial(chainAWSURI)
+		Expect(err).Should(BeNil())
+		defer client.Close()
+
+		// Loop until each node has advanced to >= the height of the block that emitted the warp log
+		for {
+			block, err := client.BlockByNumber(ctx, nil)
+			Expect(err).Should(BeNil())
+			if block.NumberU64() >= height {
```

Comment:
Can we add a log for the else case here? If one of the nodes stalls for whatever reason, I think it'd be helpful to be able to debug the tests.

Reply:
Adding a log in this tight loop would flood the output during bootstrapping, even if all nodes are healthy. We could add a sleep in the loop to reduce that, but IMO relying on the `Expect(err).Should(BeNil())` on the repeated client calls would illuminate any issues with nodes better than a "Waiting for node..." log message.

Reply:
We should still have better handling for the case in which a node is not indexing blocks correctly, but also not returning an error. I've added a context timeout to this function. When the timeout expires, the call to `client.BlockByNumber` should fail.

Reply:
Great call with the timeout 🙏